ptex 2.4 and cppmm updates

davvid opened the pull request.

davvid committed:
cppmm: change CPPMM_ENUM_STRIP to CPPMM_ENUM_PREFIX + CPPMM_RUSTIFY_ENUM

davvid committed:
api: update bindings to Ptex v2.4.0

davvid committed:
bind: update std_string.cpp for _GLIBCXX_USE_CXX11_ABI

Copy std_string.cpp's bindings for std::string from cppmm/test/std/bind.

davvid committed:
c-ptexture: add explicit instantiations for PtexPtr

Bind PtexPtr<T> institiations for PtexCache, PtexMetaData, and PtexTexture.

davvid commented on bind/c-ptexture.cpp:
```diff
@@ -383,15 +383,23 @@ template <class T> struct PtexPtr {
+using PtexMetaDataPtr = PtexPtr<Ptex::PtexMetaData>;
+
+//template class PtexPtr<Ptex::PtexTexture>;
+using PtexTexturePtr = PtexPtr<Ptex::PtexTexture>;
```
Heads up -- as soon as I uncomment the `template class` declarations above I get typedef redefinition errors:

```
ptex-c/./c-ptexture.h:239:56: error: typedef redefinition with different types
      ('Ptex__v2_4__PtexPtr_Ptex__v2_4__PtexTexture__t' (aka
      'Ptex__v2_4__PtexPtr_Ptex__v2_4__PtexTexture__t_s') vs
      'Ptex__v2_4__PtexPtr_Ptex__v2_4__PtexCache__t' (aka
      'Ptex__v2_4__PtexPtr_Ptex__v2_4__PtexCache__t_s'))
typedef Ptex__v2_4__PtexPtr_Ptex__v2_4__PtexTexture__t Ptex_PtexPtr_t;
                                                       ^
ptex-c/./c-ptexture.h:234:54: note: previous definition is here
typedef Ptex__v2_4__PtexPtr_Ptex__v2_4__PtexCache__t Ptex_PtexPtr_t;
                                                     ^
```

I left them commented out for now to get these improvements in. I'm probably doing something wrong there ;-)

davvid commented:
@anderslanglands  cppmm question: I manually hacked up my cppmm build to emit `find_package(Ptex REQUIRED)` and use `target_link_libraries(... PRIVATE Ptex::Ptex_dynamic)` in order to use Ptex's imported targets.

This is different than what cppmm emits by default for its autogenerated `CMakeLists.txt`.

I was able to build ptex-c with those changes, but I was a little unclear on what we're supposed to do next in order to build ptex-sys. I'll read the code in the openexr-bind project a bit to see if there are any hints there. Does the cargo part assume that we've installed globally or setup our environment accordingly in order to find ptex-c?

Oh.. back to the cppmm question -- how do you feel about a `--cmake-use-find-package` option that takes a string `target_link_libraries` argument as a parameter? When cppmm sees that flag it'll use `find_package` and it'll emit the user-supplied target_link_libraries arguments.

That seems like it would be sufficient to handle the use case here were we want to customize the generated cmakelists.txt. If you think that sounds like a good idea (or if you have other suggestions) let me know and I'll see if I extend cppmm for it.

davvid commented on bind/std_string.cpp:
```diff
@@ -18,7 +24,7 @@ class basic_string {
 } CPPMM_OPAQUEBYTES CPPMM_IGNORE_UNBOUND;
 
-using string = ::std::string;
+using string = std::string;
```
I somewhat copied this from the cppmm test suite. I'm really not sure if this is correct, but the `_GLIBCXX_USE_CXX11_ABI` changes seemed relevant as we prepare for the new ABI. Heads-up that cppmm uses `#ifdef` which IIUC is incorrect because this value is allowed to be either `0` or `1` and `ifdef` only checks for existence. If that's indeed a mistake we'll patch cppmm's test case to use `#if` as well.

I wasn't sure why some of the signatures use `::std` here and some use the local `std` namespace alias which expands to the ABI-specific std. Should all of them use the local std namespace alias? If so I'll tweak that too in a follow-up commit.

anderslanglands replied:
Yep the test case in cppmm itself is out of date, thanks for pointing that out. I've just pushed the correct version:
```c++
#if defined(_GLIBCXX_USE_CXX11_ABI)
#if _GLIBCXX_USE_CXX11_ABI
namespace std = ::std::__cxx11;
#else
namespace std = ::std;
#endif
#elif defined(_LIBCPP_VERSION)
namespace std = ::std::_LIBCPP_ABI_NAMESPACE;
#else
namespace std = ::std;
#endif
```

The reason some signiatures here use `::std` and some use `std` is that some are referencing the local namespace and some are referencing the global one. These are set correctly so don't change them, the above fix should make everything work correctly. If it doesn't for you let me know. I agree it's somewhat confusing - it's set this way because we need the type names in the parameters to expand to e.g. `::std::__cxx11::basic_string<char>`, but `string` is typedefed at the top-level `std` namespace.

luke-titley commented:
_I was able to build ptex-c with those changes, but I was a little unclear on what we're supposed to do next in order to build ptex-sys. I'll read the code in the openexr-bind project a bit to see if there are any hints there. Does the cargo part assume that we've installed globally or setup our environment accordingly in order to find ptex-c?_

Hey. So  for openexr, I manually wrote a build.rs that finds openexr and builds it for the -sys crate. I'm using environment variables (following anders lead on that), but it could be changed to use cmake find if that's how you wanted to do it for ptex.

_Oh.. back to the cppmm question -- how do you feel about a --cmake-use-find-package option that takes a string target_link_libraries argument as a parameter? When cppmm sees that flag it'll use find_package and it'll emit the user-supplied target_link_libraries arguments._

_That seems like it would be sufficient to handle the use case here were we want to customize the generated cmakelists.txt. If you think that sounds like a good idea (or if you have other suggestions) let me know and I'll see if I extend cppmm for it._

From my perspective that sounds like a good idea. The only alternative I can think of, is if we make it possible to provide a template CMakelists.txt file for cppmm to use. Might be more flexible...

anderslanglands commented:
Hi David! Thanks for kicking the tires!

Yeah I've been thinking about CMake as well - the current setup was good to get something up and running but we probably want to rely on CMake for setting up all the relevant options.

I *think* all our target libraries are using modern CMake these days? i.e. they're creating proper configs that can be found with `CMAKE_PREFIX_PATH`?

If so I guess the best idea would indeed be to have extra options to asttoc. Probably:
```bash
--find-package <package name>
--target-link-libraries <multiple> <library> <names>
```
e.g.
```bash
asstoc --find-package OpenEXR --target-link-libraries OpenEXR::OpenEXR OpenEXR::OpenEXRUtil
```
It would be really nice if we could pass the same arguments to astgen to find the headers - need to check how to do that.

Then when it comes to actually building the -sys crate (and its dependencies), then rather than specifying `OPENEXR_ROOT` like we do currently, `build.rs` should check `CMAKE_PREFIX_PATH` instead and if it's set, ignore the local build.

That still leaves us with having to specify the full list of libraries to link in `build.rs`, unless we can get that from CMake from there as well.

davvid commented:
Excellent, leveraging `CMAKE_PREFIX_PATH` sounds perfect. I'll rebase and pickup the updated glibcxx abi macros and drop the `s/::std/std/` edit I made.

anderslanglands commented:
Hi David, this PR on cppmm adds support for "modern" cmake: https://github.com/vfx-rs/cppmm/pull/96 and this PR on openexr-bind shows how I've currently setup the build.rs in order to be able to use it to handle both submodule-built and pre-installed target libraries: https://github.com/vfx-rs/openexr-bind/pull/39

davvid committed:
bind.sh: use the new modern cmake feature in asttoc

Setup bind.sh to use find_package(Ptex) and to find Ptex and
target_link_libraries(... Ptex::Ptex_dynamic) to build/link against it.

davvid commented:
Thanks for the `--fp` and `--tll` options -- they work great. I updated `bind.sh` here to leverage that new feature.

davvid committed:
.gitignore: ignore the generated ptex-sys output

davvid committed:
bind: add CPPMM_FORCE_NAME for std::string

davvid committed:
bind.sh: move hard-coded version settings into variables

davvid committed:
ptex: add test.rs to verify that the bindings are working

davvid commented:
Alrighty, thanks again for cppmm @anderslanglands. I added a `test.rs` similarly to how openexr-bind handles it (by copying it into the generated `ptex-sys/src/` directory) and both `cargo build` and `cargo test` are passing now, so I think this is good to merge as a starting point.

We don't have the generated `std_string::new()` that cppmm uses in its test examples here yet, though, but that's something that can be added in later once I see how that's getting setup.

I'll plan to merge this in this week.
